modelgauge/pipeline.py

```diff
@@ -0,0 +1,310 @@
+            thread.start()
+            self._workers.append(thread)
+        Thread(target=self._notice_complete).start()
+
+    @abstractmethod
+    def handle_item(self, item) -> Optional[Any]:
+        """
+        Takes in an item from the previous stage, returns the item for the next stage. If for your use
+        one input item produces multiple items, then don't return anything, instead calling
+        self.downstream_put for each output item.
+        """
+        pass
+
+    def _notice_complete(self):
+        for worker in self._workers:
+            self._debug(f"joining {worker}")
+            worker.join()
+        self._work_done.set()
+
+    def run(self):
+        self._debug(f"starting run")
+        while not self._upstream.done():
+            try:
+                self._debug(f"trying get")
+                item = self.upstream_get()
+                result = self.handle_item(item)
+                if result:
+                    self.downstream_put(result)
+                self.completed += 1
+                self.upstream_task_done()
+                self._debug(f"success with {item} -> {result}")
+            except queue.Empty:
+                pass  # that's cool
+                self._debug(f"empty")
+        self._debug(f"run finished")
+
+    def join(self):
+        self._debug(f"joining super")
+        super().join()
+        self._debug(f"joining threads")
+        for thread in self._workers:
+            thread.join()
+        self._debug(f"join done")
+
+
+class Sink(PipelineSegment):
+    """A pipeline segment that goes at the bottom. Doesn't produce, only consumes. Implement handle_item."""
+
+    def __init__(self):
+        super().__init__()
+        self._thread = None
+
+    def run(self):
+        self._debug(f"run starting")
+
+        self._work_done.clear()
+        while not self._upstream.done():
+            try:
+                item = self.upstream_get()
+                self._debug(f"handling {item}")
+                self.handle_item(item)
+                self.upstream_task_done()
+                self.completed += 1
+            except queue.Empty:
+                # that's cool
+                self._debug(f"get was empty")
+        self._work_done.set()
+        self._debug(f"finished run with upstream done")
+
+    def start(self):
+        self._thread = Thread(target=self.run)
+        self._thread.start()
+
+    @abstractmethod
+    def handle_item(self, item) -> None:
+        """Receives a work item from the previous stage. No need to return anything here."""
+        pass
+
+    def join(self):
+        super().join()
+        self._debug(f"joining thread {self._thread}")
+        self._thread.join()
+        self._debug(f"thread join complete")
+
+
+class Pipeline:
+    def __init__(
+        self,
+        *segments: PipelineSegment,
+        debug: bool = False,
+        progress_callback: Optional[Callable] = None,
+    ):
+        super().__init__()
```

Comment:
It's true that Object's init doesn't do anything, so it's true that it's technically optional. But I think it's generally a good habit to have, and since so many of the nearby` __init__` methods have it, I left it in here.